README.md
# flutter_redux_dev_tools

[![Build Status](https://travis-ci.org/brianegan/flutter_redux_dev_tools.svg?branch=master)](https://travis-ci.org/brianegan/flutter_redux_dev_tools) [![codecov](https://codecov.io/gh/brianegan/flutter_redux_dev_tools/branch/master/graph/badge.svg)](https://codecov.io/gh/brianegan/flutter_redux_dev_tools)

A Widget you can use to show a [Redux](https://pub.dartlang.org/packages/redux) Time Travel UI. Simply put it in a part of your UI that makes sense (Such as a Dev Tools Drawer), pass it a `DevToolsStore` and you'll be good to go!

Note: This Widget does not work with a normal [Redux](https://pub.dartlang.org/packages/redux) `Store`. It is meant to work with the [redux_dev_tools package](https://pub.dartlang.org/packages/redux_dev_tools), which provides a `DevToolsStore`. The `DevToolsStore` is a drop-in replacement for your Store during Development!

### Demo

A simple Flutter app that allows you to Increment and Decrement a counter.

![A screenshot of the Dev Tools in Action](https://gitlab.com/brianegan/flutter_redux_dev_tools/raw/master/devtools.gif)

### Usage

  1. Create a `main_dev.dart` file
  2. In this file, create a `DevToolsStore` in place of a normal redux `Store`
  3. Create a `ReduxDevTools` widget, passing through the Store. You can place this Widget wherever makes sense in your app! One good suggestion: In a "Dev Tools Drawer." This is generally the `endDrawer` in your Scaffold, and can contain different types of tools for a Dev Build of your app. 

### Example

This example paints only a broad outline of how to use the ReduxDevTools. For a complete example, see the `example` folder.

```dart
int addReducer(int state, action) => state + 1;

// Create a DevToolsStore instead of a normal Store during Development
final store = DevToolsStore<int>(
  addReducer,
  initialState: 0,
);

// Finally, create your app with a Redux Dev Tools
main() { 
  runApp(MaterialApp(
    home: Scaffold(
      endDrawer: ReduxDevTools<int>(store),
    ),
  ));
}
```

## Credits

All of this is inspired by the original [Redux Devtools](https://github.com/gaearon/redux-devtools).



CHANGELOG.md
# Changelog

## 0.5.2

  * Support flutter_redux >=0.8.2 <0.11.0

## 0.5.1

  * Define MaxLines for State and Action

## 0.5.0
  
  * Null Safety Support
    * Redux 5.0.0+
    * flutter_redux 0.8.2+
    * redux_dev_tools 0.7.0+

## 0.4.0

  * Upgrade to support Redux 4.x and redux_dev_tools 0.5.x

## 0.3.0

  * Update for Dart 2 & Redux 3.0.0, and redux_dev_tools 0.4.0 

## 0.2.0

  * Fix broken Action / State Dialogs for new version of Flutter
  * Redesign to fit better in a dev drawer
  * Add `ReduxDevToolsContainer`. See changes to your reducers in real time! If this Widget wraps your app, your app will automatically rerun all actions in the DevToolsStore through your reducers on Hot Reload.
  * Use latest version of `redux_dev_tools`

## 0.1.4

  * Move to github 

## 0.1.3

  * Document which Redux lib it works with

## 0.1.2

  * Add license to satisfy Pub.

## 0.1.1

  * Fix dependencies


## 0.1.0

  * Initial version, includes a `ReduxDevTools` Widget that allows you to time travel through the states of your `DevToolsStore`. Requires Dart redux and redux_dev_tools.


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter_test/flutter_test.dart';

import '../lib/main.dart';

void main() {
  testWidgets('Redux Dev Tools show the current action',
      (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new FlutterReduxApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsWidgets);
    expect(find.text('1'), findsNothing);

    // Tap the 'Increment' button and trigger a frame.
    await tester.tap(find.text("Increment"));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsWidgets);
  });
}


test/flutter_redux_dev_tools_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_redux_dev_tools/flutter_redux_dev_tools.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:redux_dev_tools/redux_dev_tools.dart';

void main() {
  int addReducer(int state, dynamic action) => state + 1;

  testWidgets('Should display the current state of the app',
      (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );

    await tester.pumpWidget(widget);

    expect(
      find.text('1'),
      findsOneWidget,
    );
  });

  testWidgets('Should display the current action', (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );
    final action = 'Hi';

    store.dispatch(action);
    await tester.pumpWidget(widget);

    expect(
      find.text(action),
      findsOneWidget,
    );
  });

  testWidgets(
      'Should display the slider at the latest state when an action is dispatched',
      (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );

    store.dispatch('Heyo');
    await tester.pumpWidget(widget);
    final slider = tester.firstWidget<Slider>(find.byType(Slider));

    expect(slider.value, 1.0);
  });

  testWidgets('Should move the slider to the original state and upate the UI',
      (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );
    final action = 'Yo yo';

    // Dispatch an extra action
    store.dispatch(action);

    await tester.pumpWidget(widget);

    // Find the Slider, and ensure it's on the latest state
    var sliderFinder = find.byType(Slider);
    var slider = tester.firstWidget<Slider>(sliderFinder);
    expect(slider.value, 1.0);
    expect(
      find.text('2'),
      findsOneWidget,
    );
    expect(
      find.text(action),
      findsOneWidget,
    );

    // Tap the slider at the left-most position, this should move the app state
    // back in time.
    await tester.tap(sliderFinder, pointer: 0);
    await tester.pump();

    slider = tester.firstWidget(sliderFinder);

    // Find Ensure the Slider has been set back to 0, and the Widget displays
    // the initial state and action.
    expect(slider.value, 0.0);
    expect(
      find.text(store.devToolsState.computedStates[0].toString()),
      findsOneWidget,
    );
    expect(
      find.text(store.devToolsState.stagedActions[0].toString()),
      findsOneWidget,
    );
  });

  testWidgets('Reset sets the store back to the last saved or initial state',
      (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );
    final action = 'Yep.';

    // Dispatch an extra action
    store.dispatch(action);

    await tester.pumpWidget(widget);

    // Find the Slider, and ensure it's on the latest state
    var sliderFinder = find.byType(Slider);
    var slider = tester.firstWidget<Slider>(sliderFinder);
    expect(slider.value, 1.0);
    expect(
      find.text('2'),
      findsOneWidget,
    );
    expect(
      find.text(action),
      findsOneWidget,
    );

    // Reset the Store by tapping a button
    await tester.tap(find.byKey(ReduxDevTools.resetKey));
    await tester.pump();

    // Ensure the UI is reset back to the initial state of the Store
    slider = tester.firstWidget(sliderFinder);
    expect(slider.value, 0.0);
    expect(
      find.text('1'),
      findsOneWidget,
    );
    expect(
      find.text(store.devToolsState.stagedActions[0].toString()),
      findsOneWidget,
    );
  });

  testWidgets(
      'Save collapses all actions into only the current state and action',
      (WidgetTester tester) async {
    final store = DevToolsStore(
      addReducer,
      initialState: 0,
    );
    final widget = MaterialApp(
      home: Scaffold(
        body: ReduxDevTools(store),
      ),
    );
    final action = 'Hey hey hey hey';

    // Dispatch an extra action
    store.dispatch(action);

    await tester.pumpWidget(widget);

    // Find the Slider, and ensure it's on the latest state
    var sliderFinder = find.byType(Slider);
    var slider = tester.firstWidget<Slider>(sliderFinder);
    expect(slider.value, 1.0);
    expect(
      find.text('2'),
      findsOneWidget,
    );
    expect(
      find.text(action),
      findsOneWidget,
    );

    // Reset the Store by tapping a button
    await tester.tap(find.byKey(ReduxDevTools.saveKey));
    await tester.pump();

    // Ensure the Slider is set back to the initial value, that the state was
    // the most recent, and that there is only 1 action int total.
    slider = tester.firstWidget(sliderFinder);
    expect(slider.value, 0.0);
    expect(
      find.text('2'),
      findsOneWidget,
    );
    expect(
      find.text(store.devToolsState.stagedActions[0].toString()),
      findsOneWidget,
    );
  });
}


example/lib/app.dart
import 'package:flutter/material.dart' hide Actions;
import 'package:flutter_redux/flutter_redux.dart';
import 'package:redux/redux.dart';

import './store.dart';

class App extends StatelessWidget {
  final Store<int> store;
  final WidgetBuilder? devDrawerBuilder;

  App({
    Key? key,
    required this.store,
    this.devDrawerBuilder,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final title = 'Flutter Redux Demo';

    return MaterialApp(
      theme: ThemeData.dark(),
      title: title,
      home: StoreProvider(
        store: store,
        child: Scaffold(
          endDrawer: devDrawerBuilder?.call(context),
          appBar: AppBar(
            title: Text(title),
          ),
          body: Container(
            child: ListView(
              children: [
                StoreConnector<int, ViewModel>(
                  distinct: true,
                  converter: (store) => ViewModel.fromStore(store),
                  builder: (context, viewModel) {
                    return Center(
                      child: Container(
                        padding: EdgeInsets.only(
                          top: 32.0,
                          bottom: 20.0,
                        ),
                        child: Column(
                          children: <Widget>[
                            Container(
                              padding: EdgeInsets.only(bottom: 28.0),
                              child: Text(
                                'Current count',
                                style: TextStyle(
                                  fontSize: 14.0,
                                  fontWeight: FontWeight.bold,
                                ),
                              ),
                            ),
                            Text(
                              viewModel.count,
                              style: Theme.of(context).textTheme.headline4,
                            ),
                            Container(
                              margin: EdgeInsets.only(
                                top: 16.0,
                                bottom: 40.0,
                              ),
                              padding: EdgeInsets.only(bottom: 40.0),
                              decoration: BoxDecoration(
                                border: Border(
                                  bottom: BorderSide(
                                    color: Colors.grey.shade800,
                                    width: 1.0,
                                  ),
                                ),
                              ),
                              child: Row(
                                children: <Widget>[
                                  Expanded(
                                    child: MaterialButton(
                                      onPressed: viewModel.onIncrementPressed,
                                      child: Text("Increment"),
                                    ),
                                  ),
                                  Expanded(
                                    child: MaterialButton(
                                      onPressed: viewModel.onDecrementPressed,
                                      child: Text("Decrement"),
                                    ),
                                  ),
                                ],
                              ),
                            )
                          ],
                        ),
                      ),
                    );
                  },
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class ViewModel {
  final String count;
  final VoidCallback onIncrementPressed;
  final VoidCallback onDecrementPressed;

  ViewModel(
    this.count,
    this.onIncrementPressed,
    this.onDecrementPressed,
  );

  factory ViewModel.fromStore(Store<int> store) {
    return ViewModel(
      store.state.toString(),
      () => store.dispatch(Actions.Increment),
      () => store.dispatch(Actions.Decrement),
    );
  }
}


example/lib/main_dev.dart
import 'package:flutter/material.dart';
import 'package:flutter_redux_dev_tools/flutter_redux_dev_tools.dart';
import 'package:redux_dev_tools/redux_dev_tools.dart';

import 'app.dart';
import 'store.dart';

// The Dev version of your app. It will build a DevToolsStore instead of a
// normal Store. In addition, it will provide a DevDrawer for the app, which
// will contain the ReduxDevTools themselves.
void main() {
  final store = DevToolsStore(
    counterReducer,
    initialState: 0,
  );

  // A ReduxDevToolsApp will recompute the state of your app on Hot Reload.
  // This means if you change the way a reducer works, it will replay all the
  // actions through the reducer to recompute the new state!
  runApp(ReduxDevToolsContainer(
    store: store,
    child: App(
      store: store,
      // Since we want a Dev Drawer that includes the Redux Dev Tools, we'll
      // provide a function that returns one! In production, notice we don't
      // provide one.
      devDrawerBuilder: (context) {
        return Drawer(
          child: Padding(
            padding: EdgeInsets.only(top: 24.0),
            child: ReduxDevTools(store),
          ),
        );
      },
    ),
  ));
}


example/.dart_tool/flutter_build/2eccb74173d8b55f364cb040b94026ee/flutter_assets.d
 /home/brian/lab/flutter_redux_dev_tools/example/build/flutter_assets/fonts/MaterialIcons-Regular.otf /home/brian/lab/flutter_redux_dev_tools/example/build/flutter_assets/AssetManifest.json /home/brian/lab/flutter_redux_dev_tools/example/build/flutter_assets/FontManifest.json /home/brian/lab/flutter_redux_dev_tools/example/build/flutter_assets/NOTICES /home/brian/lab/flutter_redux_dev_tools/example/build/flutter_assets/version.json:  /home/brian/lab/flutter_redux_dev_tools/example/pubspec.yaml /home/brian/lab/flutter/bin/cache/artifacts/material_fonts/MaterialIcons-Regular.otf /home/brian/.pub-cache/hosted/pub.dartlang.org/async-2.5.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/boolean_selector-2.1.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/characters-1.1.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/charcode-1.2.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/clock-1.1.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/collection-1.15.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/fake_async-1.2.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/flutter_redux-0.8.2/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/matcher-0.12.10/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/meta-1.3.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/path-1.8.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/redux-5.0.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/redux_dev_tools-0.7.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/source_span-1.8.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/stack_trace-1.10.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/stream_channel-2.1.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/string_scanner-1.1.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/term_glyph-1.2.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/test_api-0.2.19/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/typed_data-1.3.0/LICENSE /home/brian/.pub-cache/hosted/pub.dartlang.org/vector_math-2.1.0/LICENSE /home/brian/lab/flutter/bin/cache/pkg/sky_engine/LICENSE /home/brian/lab/flutter/packages/flutter/LICENSE /home/brian/lab/flutter_redux_dev_tools/LICENSE

example/.dart_tool/flutter_build/2eccb74173d8b55f364cb040b94026ee/linux_engine_sources.d
 /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/libflutter_linux_gtk.so /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/icudtl.dat /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_dart_project.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_method_channel.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_string_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_binary_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_message_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_view.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_event_channel.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_standard_method_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_method_response.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/flutter_linux.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_plugin_registry.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_method_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_binary_messenger.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_value.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_standard_message_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_json_method_codec.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_method_call.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_basic_message_channel.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_plugin_registrar.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_engine.h /home/brian/lab/flutter_redux_dev_tools/example/linux/flutter/ephemeral/flutter_linux/fl_json_message_codec.h:  /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/libflutter_linux_gtk.so /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/icudtl.dat /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_dart_project.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_method_channel.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_string_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_binary_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_message_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_view.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_event_channel.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_standard_method_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_method_response.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/flutter_linux.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_plugin_registry.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_method_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_binary_messenger.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_value.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_standard_message_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_json_method_codec.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_method_call.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_basic_message_channel.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_plugin_registrar.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_engine.h /home/brian/lab/flutter/bin/cache/artifacts/engine/linux-x64/flutter_linux/fl_json_message_codec.h

lib/flutter_redux_dev_tools.dart
library flutter_redux_dev_tools;

import 'package:flutter/material.dart';
import 'package:redux/redux.dart';
import 'package:redux_dev_tools/redux_dev_tools.dart';

/// A Widget you can use to show a Time Travel UI. Simply put it in a part of
/// your UI that makes sense (Such as a Dev Tools Drawer), pass it a
/// DevToolsStore and you'll be good to go!
///
/// Note: This Widget does not work with a normal Redux Store. It is meant to
/// work with the redux_dev_tools package, which provides a DevToolsStore. The
/// DevToolsStore is a drop-in replacement for your Store during Development!
///
/// ### Example
///
/// This example paints only a broad outline of how to use the ReduxDevTools.
/// For a complete example, see the `example` folder.
///
///     int addReducer(int state, action) => state + 1;
///
///     // Create a DevToolsStore instead of a normal Store during Development
///     final store = DevToolsStore(
///       addReducer,
///       initialState: 0,
///     );
///
///     // Finally, create your app:
///     runApp(MaterialApp(
///       home: Scaffold(
///         body: ReduxDevTools(store),
///       ),
///     ));
class ReduxDevTools<AppState> extends StatefulWidget {
  ReduxDevTools(this.store, {this.actionMaxLines = 1, this.stateMaxLines = 1});

  static final saveKey = UniqueKey();
  static final resetKey = UniqueKey();
  static final recomputeKey = UniqueKey();

  /// The DevToolsStore the ReduxDevTools will interact with
  final DevToolsStore<AppState> store;

  /// The number of maxLines to render action objects
  final int actionMaxLines;

  /// The number of maxLines to render state objects
  final int stateMaxLines;

  @override
  State<StatefulWidget> createState() {
    return _ReduxDevToolsState<AppState>();
  }
}

class _ReduxDevToolsState<AppState> extends State<ReduxDevTools<AppState>> {
  @override
  Widget build(BuildContext context) {
    return StreamBuilder<ReduxDevToolsViewModel>(
      stream: widget.store.onChange.map((_) {
        return ReduxDevToolsViewModel(
          widget.store,
          _ContainerState.of(context),
          context,
        );
      }),
      builder: (context, snapshot) {
        final model = snapshot.hasData
            ? snapshot.data!
            : ReduxDevToolsViewModel(
                widget.store,
                _ContainerState.of(context),
                context,
              );

        return ListView(
          children: <Widget>[
            Center(
              child: Text(
                'Redux Time Travel',
                style: TextStyle(
                  fontSize: 16.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ),
            Container(
              padding: EdgeInsets.only(
                top: 20.0,
                bottom: 12.0,
              ),
              child: Slider(
                value: model.sliderPosition,
                onChanged: model.onSliderChanged,
                min: 0.0,
                max: model.sliderMax,
              ),
            ),
            Container(
              child: Row(
                children: [
                  Expanded(
                    child: IconButton(
                      icon: Icon(Icons.save),
                      key: ReduxDevTools.saveKey,
                      onPressed: model.onSavePressed,
                      tooltip: 'Save',
                    ),
                  ),
                  Expanded(
                    child: IconButton(
                      icon: Icon(Icons.restore),
                      key: ReduxDevTools.resetKey,
                      onPressed: model.onResetPressed,
                      tooltip: 'Reset',
                    ),
                  ),
                  Expanded(
                    child: IconButton(
                      icon: Icon(
                        Icons.functions,
                        color: model.recomputeColor,
                      ),
                      key: ReduxDevTools.recomputeKey,
                      onPressed: model.onRecomputePressed,
                      tooltip: model.recomputeButtonString,
                    ),
                  ),
                ],
              ),
            ),
            Container(
              margin: EdgeInsets.symmetric(vertical: 16.0),
              child: InkWell(
                onTap: () {
                  showDialog<void>(
                    context: context,
                    builder: (context) {
                      return AlertDialog(
                        content: Container(
                          width: MediaQuery.of(context).size.width,
                          height: MediaQuery.of(context).size.height,
                          child: ListView(
                            children: [
                              Text(
                                model.latestState,
                              )
                            ],
                          ),
                        ),
                      );
                    },
                  );
                },
                child: Column(
                  children: <Widget>[
                    Container(
                      alignment: FractionalOffset.topLeft,
                      padding: EdgeInsets.only(
                        left: 12.0,
                        top: 8.0,
                        bottom: 0.0,
                        right: 0.0,
                      ),
                      child: Text(
                        'Current State',
                        style: TextStyle(
                          fontWeight: FontWeight.bold,
                          height: 1.2,
                        ),
                      ),
                    ),
                    Container(
                      alignment: FractionalOffset.topLeft,
                      padding: EdgeInsets.all(12.0),
                      child: Text(
                        model.latestState,
                        maxLines: widget.stateMaxLines,
                        overflow: TextOverflow.ellipsis,
                      ),
                    ),
                  ],
                ),
              ),
            ),
            InkWell(
              onTap: () {
                showDialog<void>(
                  context: context,
                  builder: (context) {
                    return AlertDialog(
                      content: Container(
                        width: MediaQuery.of(context).size.width,
                        height: MediaQuery.of(context).size.height,
                        child: ListView(
                          children: [
                            Text(
                              model.latestAction,
                              textAlign: TextAlign.left,
                            )
                          ],
                        ),
                      ),
                    );
                  },
                );
              },
              child: Column(
                children: <Widget>[
                  Container(
                    alignment: FractionalOffset.topLeft,
                    padding: EdgeInsets.only(
                      left: 12.0,
                      top: 8.0,
                      bottom: 0.0,
                      right: 0.0,
                    ),
                    child: Text(
                      'Current Action',
                      style: TextStyle(
                        fontWeight: FontWeight.bold,
                        height: 1.2,
                      ),
                    ),
                  ),
                  Container(
                    alignment: FractionalOffset.topLeft,
                    padding: EdgeInsets.all(12.0),
                    child: Text(
                      model.latestAction,
                      maxLines: widget.actionMaxLines,
                      overflow: TextOverflow.ellipsis,
                    ),
                  ),
                ],
              ),
            ),
          ],
        );
      },
    );
  }
}

class ReduxDevToolsViewModel {
  final String latestAction;
  final String latestState;
  final VoidCallback onSavePressed;
  final VoidCallback onResetPressed;
  final VoidCallback onRecomputePressed;
  final ValueChanged<double> onSliderChanged;
  final double sliderMax;
  final double sliderPosition;
  final Color? recomputeColor;
  final String recomputeButtonString;

  ReduxDevToolsViewModel._({
    required this.recomputeColor,
    required this.latestAction,
    required this.latestState,
    required this.onSavePressed,
    required this.onResetPressed,
    required this.onRecomputePressed,
    required this.onSliderChanged,
    required this.sliderMax,
    required this.sliderPosition,
    required this.recomputeButtonString,
  });

  factory ReduxDevToolsViewModel(
    DevToolsStore<dynamic> store,
    _ContainerState? containerState,
    BuildContext context,
  ) {
    return ReduxDevToolsViewModel._(
      latestAction: store.devToolsState.latestAction.toString(),
      latestState: store.state.toString(),
      onSavePressed: () => store.dispatch(DevToolsAction.save()),
      onResetPressed: () => store.dispatch(DevToolsAction.reset()),
      recomputeColor:
          containerState != null && containerState.recomputeOnHotReload
              ? Theme.of(context).colorScheme.secondary
              : Theme.of(context).textTheme.button?.color,
      onRecomputePressed: () {
        if (containerState != null) {
          containerState.toggleRecomputeOnHotReload();
        } else {
          store.dispatch(DevToolsAction.recompute());
        }
      },
      recomputeButtonString: containerState == null
          ? 'Recompute'
          : 'Toggle "Recompute on Hot Reload"',
      onSliderChanged: (val) =>
          store.dispatch(DevToolsAction.jumpToState(val.floor())),
      sliderMax: (store.devToolsState.computedStates.length - 1).toDouble(),
      sliderPosition: store.devToolsState.currentPosition.toDouble(),
    );
  }
}

/// Hot Reload for your State! Change your Reducers? The state will be
/// recomputed and your UI will update, pronto.
///
/// To make this work, it's best to wrap your whole app with this Widget. That
/// way it will always remain mounted and can listen to Hot Reload changes for
/// any Route.
///
/// Since you won't want to wrap your Production app with any Dev Tools, it's
/// best to only use this in Dev mode. See the `example` directory of this
/// repo for a demonstration of how to do this.
class ReduxDevToolsContainer<S> extends StatefulWidget {
  final Store<S> store;
  final Widget child;
  final bool recomputeOnHotReload;

  ReduxDevToolsContainer({
    Key? key,
    required this.store,
    required this.child,
    this.recomputeOnHotReload = true,
  }) : super(key: key);

  @override
  _ReduxDevToolsRecomputeState createState() {
    return _ReduxDevToolsRecomputeState();
  }
}

class _ReduxDevToolsRecomputeState extends State<ReduxDevToolsContainer> {
  bool _recomputeOnHotReload = false;

  @override
  Widget build(BuildContext context) {
    return _ContainerState(
      recomputeOnHotReload: _recomputeOnHotReload,
      toggleRecomputeOnHotReload: _toggleRecomputeOnHotReload,
      child: widget.child,
    );
  }

  void _toggleRecomputeOnHotReload() {
    setState(() {
      _recomputeOnHotReload = !_recomputeOnHotReload;
    });
  }

  @override
  void reassemble() {
    if (_recomputeOnHotReload) {
      widget.store.dispatch(DevToolsAction.recompute());
    }

    super.reassemble();
  }

  @override
  void initState() {
    super.initState();

    _recomputeOnHotReload = widget.recomputeOnHotReload;
  }
}

class _ContainerState extends InheritedWidget {
  final bool recomputeOnHotReload;
  final Function() toggleRecomputeOnHotReload;

  _ContainerState({
    Key? key,
    required Widget child,
    required this.recomputeOnHotReload,
    required this.toggleRecomputeOnHotReload,
  }) : super(key: key, child: child);

  static _ContainerState? of(BuildContext context) {
    return context.dependOnInheritedWidgetOfExactType<_ContainerState>();
  }

  @override
  bool updateShouldNotify(_ContainerState oldWidget) {
    return recomputeOnHotReload != oldWidget.recomputeOnHotReload;
  }
}
